### PR TITLE
feat(agw): restart services on shared config changes

### PR DESCRIPTION
**Pull request by ankitMeta1**
## Summary

while processing updates to gateway.mconfig, we check if shared_mconfig changes, if yes, add all the services to restart list.

## Test Plan

ran various scenarios on local vm.

Screen Shot 2021-12-10 at 1 47 56 PM
Screen Shot 2021-12-10 at 1 49 09 PM
Screen Shot 2021-12-10 at 1 47 46 PM

### CODE REVIEW

File: orc8r/gateway/python/magma/magmad/config_manager.py [hunk 1]
```diff
@@ -118,10 +118,15 @@ def did_mconfig_change(serv_name):
                 ),
             )
 
-        # TODO adapt service restart logic to include changes in shared_mconfig
```
Comment (themarwhal) [inline orc8r/gateway/python/magma/magmad/config_manager.py:121]: Let's also restart magmad if we modify the shared config as well ^ see line 113

**Comment by themarwhal** [comment]
Rename PR title to "feat(agw): restart services on shared config changes" to make the semantic PR check happy :)

File: orc8r/gateway/python/magma/magmad/config_manager.py [hunk 2]
```diff
@@ -118,10 +118,15 @@ def did_mconfig_change(serv_name):
...
-        services_to_restart = [
-            srv for srv in self._services if did_mconfig_change(srv)
-        ]
+        services_to_restart = []
+        if 'shared_mconfig' in mconfig.configs_by_key and did_mconfig_change('shared_mconfig'):
```
Comment (themarwhal) [inline orc8r/gateway/python/magma/magmad/config_manager.py]: let's define a const value for this so we avoid typos :D

File: orc8r/gateway/python/magma/magmad/config_manager.py [hunk 3]
```diff
@@ -118,10 +118,15 @@ def did_mconfig_change(serv_name):
...
+            logging.info("shared config changed. Restarting all services.")
```
Comment (themarwhal) [inline orc8r/gateway/python/magma/magmad/config_manager.py]:
```suggestion
            logging.info("Shared config changed. Restarting all services.")
```

**Review by themarwhal — APPROVED** [review]
LGTM :)

File: orc8r/gateway/python/magma/magmad/config_manager.py [hunk 4]
```diff
@@ -109,19 +110,31 @@ def did_mconfig_change(serv_name):
             return mconfig.configs_by_key.get(serv_name) != \
                 self._mconfig.configs_by_key.get(serv_name)
 
-        # Reload magmad configs locally
-        if did_mconfig_change('magmad'):
+        def restart_magmad():
```
Comment (themarwhal) [inline orc8r/gateway/python/magma/magmad/config_manager.py]: Does this function restart magmad or only the dynamic servicecs loaded from the magmad config?

File: orc8r/gateway/python/magma/magmad/config_manager.py [hunk 5]
```diff
@@ -109,19 +110,31 @@ def did_mconfig_change(serv_name):
...
                 ),
             )
 
-        # TODO adapt service restart logic to include changes in shared_mconfig
-        services_to_restart = [
-            srv for srv in self._services if did_mconfig_change(srv)
-        ]
+        magmad_restarted = False
+        # Reload magmad configs locally
+        if did_mconfig_change('magmad'):
```
Comment (themarwhal) [inline orc8r/gateway/python/magma/magmad/config_manager.py]: might be cleaner to keep these two logics separate. So something like 
```python
if shared_mconfig_changed or magmad_changed:
   restart things

# process other services
```

**Review by themarwhal — CHANGES_REQUESTED** [review]
looks like the lte-test check is failing on CI. The GitHub action checks use a different env than the magma VM so maybe something is missing there: https://github.com/magma/magma/blob/master/.github/workflows/agw-workflow.yml#L52

File: orc8r/gateway/python/magma/magmad/config_manager.py [hunk 6]
```diff
@@ -82,6 +82,7 @@ def process_update(self, stream_name, updates, resync):
                     since config is contained in one DB element, hence all
                     data is sent in every update.
         """
+        shared_mconfig = 'shared_mconfig'
```
Comment (themarwhal) [inline orc8r/gateway/python/magma/magmad/config_manager.py]: Let's move this up to the top of the file as other consts are defined that way.
See https://github.com/magma/magma/blob/master/orc8r/gateway/python/magma/magmad/config_manager.py#L29

**Comment by themarwhal** [comment]
Would be great if you could file an issue for removing the config dependency for lte-test and take that on. :)

File: orc8r/gateway/python/magma/magmad/config_manager.py [hunk 7]
```diff
@@ -110,18 +112,27 @@ def did_mconfig_change(serv_name):
                 self._mconfig.configs_by_key.get(serv_name)
 
         # Reload magmad configs locally
-        if did_mconfig_change('magmad'):
+        if did_mconfig_change(MAGMAD) or (
+            SHARED_MCONFIG in mconfig.configs_by_key
+            and did_mconfig_change(SHARED_MCONFIG)
+        ):
+            logging.info("Restarting dynamic Services.")
```
Comment (themarwhal) [inline orc8r/gateway/python/magma/magmad/config_manager.py]:
```suggestion
            logging.info("Restarting dynamic services due to config change")
```

**Comment by ankitMeta1** [comment]
> Would be great if you could file an issue for removing the config dependency for lte-test and take that on. :)

I would create a issue and add it to the backlog, and try to pick it up later :)

**Pull request closed without merging**